### ISSUE TITLE
random_mod can take a very long time

### ISSUE THREAD

**Issue by Sc00bz** [RustCrypto/utils#618]
https://github.com/RustCrypto/utils/blob/710abc8c83a13def1c01958909405976f915e5a5/crypto-bigint/src/uint/rand.rs#L34-L42

If `modulus` is small compared to the full size ("number of limbs * size of limb") then it can take "forever".

The correct method is to mask off the top bits of the random number to match the size of `modulus`. This leads to a worst case of a 50% chance of failing and needing to generate another random number. The average for this worst case is needing to generate 2 random numbers.

**Comment by Sc00bz** [comment]
Oh right obviously this can run forever if called with a modulus of zero. For this case it should error similarly to a divide by zero.

**Comment by tarcieri** [comment]
For now we can add a `debug_assert!` that the modulus is non-zero.

 It'd be nice to eventually capture these sort of cases [using the type system](https://blog.rust-lang.org/inside-rust/2021/09/06/Splitting-const-generics.html) so we can ensure implementations are panic free.

**Comment by newpavlov** [comment]
I think that ideally modulus should be a const parameter for a higher-level bigint type. This would allow us to either reject such edge cases on the type level or insert branches which would be eliminated at compile time.

**Comment by tarcieri** [comment]
@newpavlov yep, exactly, unfortunately `min_const_generics` are nowhere near that expressive
